Complex backward for indexing, slicing, joining, and mutating ops

anjali411 opened the pull request:
This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49552 Complex backward for indexing, slicing, joining, and mutating ops**

Differential Revision: [D25682511](https://our.internmc.facebook.com/intern/diff/D25682511)

anjali411 committed:
Complex backward for indexing, slicing, joining, and mutating ops

[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex backward for indexing, slicing, joining, and mutating ops"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex backward for indexing, slicing, joining, and mutating ops"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -281,6 +281,77 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
+            SampleInput((make_tensor((M, S), device, dtype,
+                                     low=None, high=None,
+                                     requires_grad=requires_grad),
+                        1, gather_variable((M, S // 2), 0, S, True, device=device))),
```
Unfortunately I think both test_autograd.py and this file have identical gather_variable implementations. Maybe this PR can remove the version in test_autograd.py? I think the same is true for index_variable.

anjali411 replied:
The tests here use the existing implementation of `gather_variable` and `index_variable`

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -800,6 +871,86 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+                        dtypes=[torch.complex64, torch.complex128, torch.float16, torch.bfloat16]),
+           ),
+           sample_inputs_func=sample_inputs_index_select),
+    OpInfo('stack',
```
Add a comment explaining the need for this wrapper

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -800,6 +871,86 @@ def sample_inputs(self, device, dtype, requires_grad=False):
+           ),
+           sample_inputs_func=sample_inputs_stack),
+    OpInfo('hstack',
+           op=lambda *args: torch.hstack([*args]),
```
Add a comment explaining the need for this wrapper, too

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -281,6 +281,77 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
+            SampleInput((make_tensor((), device, dtype,
+                                     low=None, high=None,
+                                     requires_grad=requires_grad),
+                        0, torch.tensor([0], dtype=torch.int64))),
```
This tensor and the one in the next sample input are always created on the CPU, causing CI failures

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -281,6 +281,77 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
+            SampleInput((make_tensor((), device, dtype,
+                                     low=None, high=None,
+                                     requires_grad=requires_grad),
+                        0, torch.tensor([0], dtype=torch.int64))),
```
This tensor and the tensors used in the next two sample inputs are always on the CPU, causing CI failures

mruberry reviewed (commented):
Hey @anjali411!

Overall this looks very good. I've made a few comments inline. Rebasing should get you past most CI failures you're seeing, but there are some real ones. I highlighted what I think is causing them with comments.

This PR is also skipping a lot of tests, which is surprising. Are those skips really necesssary? If so, we should probably file issues and/or add comments explaining them.

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.




[ghstack-poisoned]

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.




[ghstack-poisoned]

anjali411 commented:
@mruberry updated the PR based on your comments and also filed an issue for some failing tests. As discussed offline, we should file a separate issue to add `grad_op` or something on those lines to ensure jit tests are unaffected when we pass a wrapper for some reason with flattened arguments.

mruberry commented:
`test_unsupported_dtypes_gather_cuda_bfloat16 - TestOpInfoCUDA` failure is real because gather on CUDA does support bfloat16:

```
t = torch.tensor([[1,2],[3,4]], dtype=torch.bfloat16)
torch.gather(t, 1, torch.tensor([[0,0],[1,0]]))
: RuntimeError: method_name not implemented for 'BFloat16'

t = torch.tensor([[1,2],[3,4]], dtype=torch.bfloat16, device='cuda')
torch.gather(t, 1, torch.tensor([[0,0],[1,0]], device='cuda'))
: tensor([[1., 1.],
  [4., 3.]], device='cuda:0', dtype=torch.bfloat16)
```

You need to update its dtypesIfCUDA.

The ROCm error is also real in `test_sign_complex_assert_raises_cuda`. See here:

https://github.com/pytorch/pytorch/blob/159de1f1d6c19332e90f12623bf7516ea6d46592/test/test_unary_ufuncs.py#L609

I don't know why, but the test verifies index add doesn't work on complex types for some reason:

https://github.com/pytorch/pytorch/blob/159de1f1d6c19332e90f12623bf7516ea6d46592/test/test_unary_ufuncs.py#L618

Seems like that check can just be removed?

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.


Differential Revision: [D25682511](https://our.internmc.facebook.com/intern/diff/D25682511)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.


Differential Revision: [D25682511](https://our.internmc.facebook.com/intern/diff/D25682511)

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1059,6 +1130,75 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
            supports_tensor_out=False,
            sample_inputs_func=sample_inputs_pinverse,
            decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack]),
+    OpInfo('gather',
```
method_tests for gather can be removed here:

https://github.com/pytorch/pytorch/blob/97c17b47720dd1afce88180b843c9182f1a665a9/torch/testing/_internal/common_methods_invocations.py#L2087

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1059,6 +1130,75 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
+           dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+           test_inplace_grad=False,
+           sample_inputs_func=sample_inputs_gather),
+    OpInfo('index_select',
```
method_tests for index_select can be removed here:

https://github.com/pytorch/pytorch/blob/97c17b47720dd1afce88180b843c9182f1a665a9/torch/testing/_internal/common_methods_invocations.py#L1886

mruberry commented:
Several of the builds hit a timeout at the same place, here's a snippet from the ASAN build:

```
Dec 28 18:43:15   test_tile_empty_tensor_complex_cpu (__main__.TestAutogradDeviceTypeCPU) ... ok (0.980s)
Dec 28 18:43:15   test_tile_empty_tensor_cpu (__main__.TestAutogradDeviceTypeCPU) ... ok (0.257s)
Dec 28 18:44:57   test_tile_less_reps_dims_complex_cpu (__main__.TestAutogradDeviceTypeCPU) ... ok (101.370s)
Dec 28 18:45:16   test_tile_less_reps_dims_cpu (__main__.TestAutogradDeviceTypeCPU) ... ok (19.329s)

Too long with no output (exceeded 1h30m0s): context deadline exceeded
```

Which is worrying since I believe this means that the timeouts are happening in test_tile_more_reps_dims_complex_cpu, which this PR is adding. The non-complex version of this test was already incredibly slow. I filed an issue: https://github.com/pytorch/pytorch/issues/49962.

I think there are two options for how to proceed:

- changes to tile can be reverted from this PR and we can put this in as-is, following-up later to remove the redundant method tests and change tile
- an OpInfo can be added for tile and we can skip the tests that take too long

We could also looking at trying to skip the tile tests in situ but, but I think that'd be trickier.

mruberry commented:
Rethinking modifying the existing tests, decreasing the size of the tile tests to be comparable to the repeat tests also seems like an option.

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.


Differential Revision: [D25682511](https://our.internmc.facebook.com/intern/diff/D25682511)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.


Differential Revision: [D25682511](https://our.internmc.facebook.com/intern/diff/D25682511)

[ghstack-poisoned]

anjali411 committed:
Update on "Complex backward for indexing, slicing, joining, and mutating ops"

This PR:
1. Migrates independent autograd test for `hstack`, `dstack`, `vstack`, `movedim`, `moveaxis` from `test_autograd.py` to the new `OpInfo` based tests.
2. Migrates autograd test for `gather`, `index_select` from the method_tests to the new `OpInfo` based tests.
2. Enables complex backward for `stack, gather, index_select, index_add_` and adds tests for complex autograd for all the above mentioned ops.


Differential Revision: [D25682511](https://our.internmc.facebook.com/intern/diff/D25682511)

[ghstack-poisoned]

mruberry reviewed (approved):
Awesome! Thanks for fixing tile's gradcheck test performance, too.
